src/google/cloud/ndb/_datastore_query.py

```diff
@@ -96,59 +95,396 @@ def fetch(query):
-        # have to applied to the aggregate, not passed to Datastore to use on
-        # individual queries
-        offset = query.offset
-        limit = query.limit
-        query = query.copy(offset=0, limit=None)
-    else:
-        offset = limit = None
-
-    queries = [
-        _run_query(project_id, namespace, _query_to_protobuf(query, filter_pb))
-        for filter_pb in filter_pbs
-    ]
-    result_sets = yield queries
-    result_sets = [
-        (
+    results = iterate(query)
+    entities = []
+    while (yield results.has_next_async()):
+        entities.append(results.next())
+
+    return entities
+
+
+def iterate(query):
+    """Get iterator for query results.
+
+    Args:
+        query (query.QueryOptions): The query spec.
+
+    Returns:
+        QueryIterator: The iterator.
+    """
+    filters = query.filters
+    if filters and filters._multiquery:
+        return _MultiQueryIteratorImpl(query)
+
+    return _QueryIteratorImpl(query)
+
+
+class QueryIterator:
+    """An iterator for query results.
+
+    Executes the given query and provides an interface for iterating over
+    instances of either :class:`model.Model` or :class:`key.Key` depending on
+    whether ``keys_only`` was specified for the query.
+
+    This is an abstract base class. Users should not all an iterator class
+    directly. Use :meth:`query.Query.iter` or ``iter(query)`` to get an
+    instance of :class:`QueryIterator`.
+    """
+
+    def __iter__(self):
+        return self
+
+    def has_next(self):
+        """Is there at least one more result?
+
+        Blocks until the answer to this question is known and buffers the
+        result (if any) until retrieved with :meth:`next`.
+
+        Returns:
+            bool: :data:`True` if a subsequent call to
+                :meth:`QueryIterator.next` will return a result, otherwise
+                :data:`False`.
+        """
+        raise NotImplementedError()
+
+    def has_next_async(self):
+        """Asynchronous version of :meth:`has_next`.
+
+        Returns:
+            tasklets.Future: See :meth:`has_next`.
+        """
+        raise NotImplementedError()
+
+    def probably_has_next(self):
+        """Like :meth:`has_next` but won't block.
+
+        This uses a (sometimes inaccurate) shortcut to avoid having to hit the
+        Datastore for the answer.
+
+        May return a false positive (:data:`True` when :meth:`next` would
+        actually raise ``StopIteration``), but never a false negative
+        (:data:`False` when :meth:`next` would actually return a result).
+        """
+        raise NotImplementedError()
+
+    def next(self):
+        """Get the next result.
+
+        May block. Guaranteed not to block if immediately following a call to
+        :meth:`has_next` or :meth:`has_next_async` which will buffer the next
+        result.
+
+        Returns:
+            Union[model.Model, key.Key]: Depending on if the '`keys_only`'
```

Comment:
Depending on if the '`keys_only`' option is being used? Or just remove the "the" before keys_only.